cns/requestcontroller/kubecontroller/crdrequestcontroller.go

```diff
@@ -0,0 +1,168 @@
+
+	//Check that logger package has been intialized
+	if logger.Log == nil {
+		return nil, errors.New("Must initialize logger before calling")
+	}
+
+	// Check that NODENAME environment variable is set. NODENAME is name of node running this program
+	nodeName := os.Getenv(nodeNameEnvVar)
+	if nodeName == "" {
+		return nil, errors.New("Must declare " + nodeNameEnvVar + " environment variable.")
+	}
+
+	//Add client-go scheme to runtime sheme so manager can recognize it
+	var scheme = runtime.NewScheme()
+	if err := clientgoscheme.AddToScheme(scheme); err != nil {
+		return nil, errors.New("Error adding client-go scheme to runtime scheme")
+	}
+
+	//Add CRD scheme to runtime sheme so manager can recognize it
+	if err := nnc.AddToScheme(scheme); err != nil {
+		return nil, errors.New("Error adding NodeNetworkConfig scheme to runtime scheme")
+	}
+
+	// Create manager for CrdRequestController
+	// MetricsBindAddress is the tcp address that the controller should bind to
+	// for serving prometheus metrics, set to "0" to disable
+	mgr, err := ctrl.NewManager(kubeconfig, ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: prometheusAddress,
+		Namespace:          k8sNamespace,
+	})
+	if err != nil {
+		logger.Errorf("[cns-rc] Error creating new request controller manager: %v", err)
+		return nil, err
+	}
+
+	//Create cnsinteractor
+	cnsInteractor := &CNSInteractor{
+		RestService: restService,
+	}
+
+	//Create reconciler
+	crdreconciler := &CrdReconciler{
+		APIClient: mgr.GetClient(),
+		NodeName:  nodeName,
+		CNSClient: cnsInteractor,
+	}
+
+	// Setup manager with reconciler
+	if err := crdreconciler.SetupWithManager(mgr); err != nil {
+		logger.Errorf("[cns-rc] Error creating new CrdRequestController: %v", err)
+		return nil, err
+	}
+
+	// Create the requestController
+	crdRequestController := crdRequestController{
+		mgr:        mgr,
+		APIClient:  mgr.GetClient(),
+		nodeName:   nodeName,
+		Reconciler: crdreconciler,
+	}
+
+	return &crdRequestController, nil
+}
+
+// StartRequestController starts the Reconciler loop which watches for CRD status updates
+// Blocks until SIGINT or SIGTERM is received
+// Notifies exitChan when kill signal received
+func (crdRC *crdRequestController) StartRequestController(exitChan chan bool) error {
+	logger.Printf("Starting manager")
+	if err := crdRC.mgr.Start(SetupSignalHandler(exitChan)); err != nil {
+		logger.Errorf("[cns-rc] Error starting manager: %v", err)
+	}
+
+	return nil
+}
+
+// UpdateCRDSpec updates the CRD spec
+func (crdRC *crdRequestController) UpdateCRDSpec(cntxt context.Context, crdSpec *nnc.NodeNetworkConfigSpec) error {
+	nodeNetworkConfig, err := crdRC.getNodeNetConfig(cntxt, crdRC.nodeName, k8sNamespace)
+	if err != nil {
+		logger.Errorf("[cns-rc] Error getting CRD when updating spec %v", err)
+		return err
+	}
+
+	//Update the CRD spec
+	nodeNetworkConfig.Spec = *crdSpec
```

Comment:
I think you should use `crdspec.DeepCopyInto(&nodeNetworkConfig.Spec)` here for a safer assignment https://github.com/Azure/azure-container-networking/blob/master/nodenetworkconfig/api/v1alpha/zz_generated.deepcopy.go#L122